File: built-in plugins dsp/mb-supersaw-lead.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSupersawLeadPlugin: PluginDefinition = {
    id: 'mb-supersaw-lead',
    slug: 'mb-supersaw-lead',
    name: 'MB Supersaw Lead',
    category: 'instrument',
    type: 'synth',
    version: '1.0.0',
    description: 'Massive supersaw lead synth for EDM and trance productions',
    author: 'Max Booster',
    oscillators: [
      { type: 'sawtooth', detune: -25, gain: 0.15 },
      { type: 'sawtooth', detune: -15, gain: 0.2 },
      { type: 'sawtooth', detune: -5, gain: 0.25 },
      { type: 'sawtooth', detune: 0, gain: 0.3 },
      { type: 'sawtooth', detune: 5, gain: 0.25 },
      { type: 'sawtooth', detune: 15, gain: 0.2 },
      { type: 'sawtooth', detune: 25, gain: 0.15 },
    ],
    envelope: { attack: 0.01, decay: 0.2, sustain: 0.8, release: 0.4 },
    parameters: [
      { id: 'voices', name: 'Supersaw Voices', type: 'int', defaultValue: 7, minValue: 3, maxValue: 16, automatable: false },
      { id: 'detune', name: 'Supersaw Detune', type: 'float', defaultValue: 25, minValue: 0, maxValue: 100, unit: 'cents', automatable: true },
      { id: 'filter_cutoff', name: 'Filter Cutoff', type: 'float', defaultValue: 12000, minValue: 100, maxValue: 20000, unit: 'Hz', automatable: true },
      { id: 'filter_res', name: 'Filter Resonance', type: 'float', defaultValue: 0.2, minValue: 0, maxValue: 1, automatable: true },
      { id: 'stereo_width', name: 'Stereo Width', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
      { id: 'attack', name: 'Attack', type: 'float', defaultValue: 0.01, minValue: 0.001, maxValue: 2, unit: 's', automatable: true },
      { id: 'release', name: 'Release', type: 'float', defaultValue: 0.4, minValue: 0.01, maxValue: 5, unit: 's', automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { voices: 7, detune: 25, filter_cutoff: 12000, filter_res: 0.2, stereo_width: 0.8, attack: 0.01, release: 0.4, volume: 0.8 },
  };

export default MbSupersawLeadPlugin;

File: built-in plugins dsp/mb-supersaw-lead.cpp
/**
 * MB Supersaw Lead
 * Category : instrument
 * Type     : synth
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Massive supersaw lead synth for EDM and trance productions
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SUPERSAW_LEAD_H
#define MB_SUPERSAW_LEAD_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSupersawLead : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-supersaw-lead";
    static constexpr const char* PLUGIN_NAME    = "MB Supersaw Lead";
    static constexpr const char* PLUGIN_TYPE    = "synth";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float voices = 7f;  // range [3, 16]
    float detune = 25f;  // range [0, 100]
    float filter_cutoff = 12000f;  // range [100, 20000]
    float filter_res = 0.2f;  // range [0, 1]
    float stereo_width = 0.8f;  // range [0, 1]
    float attack = 0.01f;  // range [0.001, 2]
    float release = 0.4f;  // range [0.01, 5]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSupersawLead() = default;
    ~MbSupersawLead() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.voices = std::clamp(params.voices, 3f, 16f);
        params.detune = std::clamp(params.detune, 0f, 100f);
        params.filter_cutoff = std::clamp(params.filter_cutoff, 100f, 20000f);
        params.filter_res = std::clamp(params.filter_res, 0f, 1f);
        params.stereo_width = std::clamp(params.stereo_width, 0f, 1f);
        params.attack = std::clamp(params.attack, 0.001f, 2f);
        params.release = std::clamp(params.release, 0.01f, 5f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Supersaw Lead
        return input;
    }
};

#endif // MB_SUPERSAW_LEAD_H
